src/main/java/fi/aalto/cs/apluscourses/model/MainViewModelUpdater.java

```diff
@@ -0,0 +1,172 @@
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class MainViewModelUpdater {
+
+  @NotNull
+  private MainViewModelProvider mainViewModelProvider;
+
+  @NotNull
+  private APlusProject aplusProject;
+
+  private long updateInterval;
+
+  /**
+   * Construct a {@link MainViewModelUpdater} with the given {@link MainViewModelProvider}, project,
+   * and update interval.
+   * @param mainViewModelProvider The {@link MainViewModelProvider} from which the main view model
+   *                              to be updated is gotten.
+   * @param project               The project to which this updater is tied.
+   * @param updateInterval        The interval at which the updater performs the update.
+   */
+  public MainViewModelUpdater(@NotNull MainViewModelProvider mainViewModelProvider,
+                              @NotNull Project project,
+                              @NotNull long updateInterval) {
+    this.mainViewModelProvider = mainViewModelProvider;
+    this.aplusProject = new APlusProject(project);
+    this.updateInterval = updateInterval;
+  }
+
+  /**
+   * Construct a {@link MainViewModelUpdater} with the given project.
+   * @param project The project to which thsi updater is tied.
+   */
+  public MainViewModelUpdater(@NotNull Project project) {
+    this(PluginSettings.getInstance(), project, PluginSettings.MAIN_VIEW_MODEL_UPDATE_INTERVAL);
+  }
+
+  private void interruptIfProjectClosed() throws InterruptedException {
+    if (aplusProject.getProject().isDisposed() || !aplusProject.getProject().isOpen()) {
+      throw new InterruptedException();
+    }
+  }
+
+  @NotNull
+  URL getCourseUrl() throws InterruptedException {
+    return ReadAction.compute(() -> {
+      interruptIfProjectClosed();
+      try {
+        URL url = aplusProject.getCourseFileUrl();
+        if (url == null) {
+          throw new InterruptedException();
+        }
+        return url;
+      } catch (IOException e) {
+        throw new InterruptedException();
+      }
+    });
+  }
+
+  @Nullable
+  Course getCourse(@NotNull URL courseUrl) {
+    // We don't need to be inside the read action when we parse the course configuration file.
+    // This way we avoid long read actions, which block the UI thread from making write
+    // actions and lead to UI freezes. We just have to remember to later check again that the
+    // project is still actually open.
+    try {
+      return Course.fromUrl(courseUrl, new IntelliJModelFactory(aplusProject.getProject()));
+    } catch (IOException | MalformedCourseConfigurationFileException e) {
+      return null;
+    }
+  }
+
+  @NotNull
+  Set<String> getProjectModules() throws InterruptedException {
```

Comment:
I do not thoroughly understand, but could those methods (previous and the next one as well) be located elsewhere (like in the APlusProject class)?